### PR TITLE
8426 add a config option so collectd can generate multiple fields, not multiple measurements

### PR DESCRIPTION
**Pull request by aanthony1243**
###### Required for all non-trivial PRs
- [X] Rebased/mergable
- [X] Tests pass
- [X] CHANGELOG.md updated
- [X] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

###### Required only if applicable
_You can erase any checkboxes below this note if they are not applicable to your Pull Request._
- [X] Config changes: update sample config (`etc/config.sample.toml`), server `NewDemoConfig` method, and `Diagnostics` methods reporting config settings, if necessary
- [X] [InfluxData Documentation](https://github.com/influxdata/docs.influxdata.com): issue filed or pull request submitted <Doc updated implicitly via new Readme.md on this PR>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,15 @@
 ## v1.4.0 [unreleased]
 
+### Configuration Changes
+
+#### `[collectd]` Section
+
+* `parse-multivalue-plugin` was added with a default of `split`.  When set to `split`, multivalue plugin data will be split into separate measurements.  When set to `join`, multivalue plugin will be stored as a single multi-value measurement.
```
Comment (jwilder) [inline CHANGELOG.md]: This config option needs to be added to the sample config in `etc/config.sample.toml` too.

File: services/collectd/service.go [hunk 2]
```diff
@@ -385,6 +390,53 @@ func (s *Service) writePoints() {
...
+// The difference is that the original provided measurements like (PLUGIN_DSNAME, ["value",xxx])
+// while this one will provide measurements like (PLUGIN, {["DSNAME",xxx]})
+// this effectively joins collectd data that should go together, such as:
+//  (df, {["used",1000],["free",2500]})
+func (s *Service) UnmarshalValueListPacked(vl *api.ValueList) []models.Point {
+	timestamp := vl.Time.UTC()
+
+	var name = vl.Identifier.Plugin
+	tags := make(map[string]string)
+	fields := make(map[string]interface{})
```
Comment (mark-rushakoff) [inline services/collectd/service.go]: Although Go allows you to `make` maps (and slices) without specifying a size, but we've run into many bottlenecks that were (at least partially) addressed by specifying an initial size hint. 

Here, I'd go with 
```
tags := make(map[string]string, 4)
fields := make(map[string]interface{}, len(vl.Values))
```
We might not use all 4 `tags` slots, which is fine, but giving `fields` the right capacity from the start will avoid creating some garbage as the map is grown from a smaller capacity.

File: services/collectd/service.go [hunk 3]
```diff
@@ -385,6 +390,53 @@ func (s *Service) writePoints() {
...
+	if vl.Identifier.Type != "" {
+		tags["type"] = vl.Identifier.Type
+	}
+	if vl.Identifier.TypeInstance != "" {
+		tags["type_instance"] = vl.Identifier.TypeInstance
+	}
+
+	for i := range vl.Values {
+		fieldName := vl.DSName(i)
+		switch value := vl.Values[i].(type) {
```
Comment (mark-rushakoff) [inline services/collectd/service.go]: It's generally preferable to use the two-argument range function rather than indexing again, i.e.

```
for i, v := range vl.Values {
  fieldName := vl.DSName(i)
  switch val := v.(type) {
```

File: services/collectd/service.go [hunk 4]
```diff
@@ -385,6 +390,53 @@ func (s *Service) writePoints() {
...
+		case api.Counter:
+			fields[fieldName] = float64(value)
+		}
+	}
+	// Drop invalid points
+	p, err := models.NewPoint(name, models.NewTags(tags), fields, timestamp)
+	if err != nil {
+		s.Logger.Info(fmt.Sprintf("Dropping point %v: %v", name, err))
+		atomic.AddInt64(&s.stats.InvalidDroppedPoints, 1)
+	}
```
Comment (mark-rushakoff) [inline services/collectd/service.go]: If `err` is not nil, we may not be sure of the state of `p`. This block continues to the return statement. Should it `return nil` instead?

File: services/collectd/service.go [hunk 5]
```diff
@@ -385,6 +390,53 @@ func (s *Service) writePoints() {
...
+
+	// return as an array for compatability with rest of API
```
Comment (mark-rushakoff) [inline services/collectd/service.go]: Probably not a necessary comment, and it's returning a slice, not an array. https://blog.golang.org/go-slices-usage-and-internals

File: services/collectd/service_test.go [hunk 6]
```diff
@@ -260,13 +260,73 @@ func TestService_BatchSize(t *testing.T) {
 	}
 }
 
+// Test that the collectd service correctly batches points by BatchSize.
```
Comment (mark-rushakoff) [inline services/collectd/service_test.go]: Looks like a copy-pasted comment.

File: services/collectd/service_test.go [hunk 7]
```diff
@@ -260,13 +260,73 @@ func TestService_BatchSize(t *testing.T) {
...
+	}
+
+	// Send the test data to the service.
+	if n, err := conn.Write(testData); err != nil {
+		t.Fatal(err)
+	} else if n != len(testData) {
+		t.Fatalf("only sent %d of %d bytes", n, len(testData))
+	}
+
+	points := []models.Point{}
```
Comment (mark-rushakoff) [inline services/collectd/service_test.go]: Since you're only appending to this, it's okay to write `var points []models.Point`. Go will do the right thing when you say `points = append(points, foo)`.

File: services/collectd/service_test.go [hunk 8]
```diff
@@ -260,13 +260,73 @@ func TestService_BatchSize(t *testing.T) {
...
+
+Loop:
+	for {
+		select {
+		case p := <-pointCh:
+			points = append(points, p)
+			if len(points) == totalPoints {
+				break Loop
+			}
+		case <-time.After(time.Second):
```
Comment (mark-rushakoff) [inline services/collectd/service_test.go]: Use of `time.After` in a  loop like this is an antipattern. 

Every time the select block is entered, a new timer is created. The timer continues to run in the background for the full second, and there's overhead in the runtime for managing any running timer.

It's better to use a single instance of [`time.Timer`](https://golang.org/pkg/time/#Timer) and call `Reset` on each iteration of the loop.

File: CHANGELOG.md (hunk 1 above)
Comment (aanthony1243) [inline CHANGELOG.md]: forgot to push...it's there now.

Comment (mark-rushakoff) [inline CHANGELOG.md]: It's hard to follow what a multivalue plugin means exactly. Is this something that collectd users are expected to know? If not, it would be helpful to have an example of the differences between split and join.

File: services/collectd/service.go (hunk 2 above)
Comment (aanthony1243) [inline services/collectd/service.go]: This snippet was pulled from another test, should I fix the other tests too?

File: services/collectd/service_test.go (hunk 7 above)
Comment (aanthony1243) [inline services/collectd/service_test.go]: done.

File: services/collectd/service.go (hunk 2 above)
Comment (aanthony1243) [inline services/collectd/service.go]: done.

File: services/collectd/service.go [hunk 9]
```diff
@@ -385,6 +390,53 @@ func (s *Service) writePoints() {
 	}
 }
 
+// UnmarshalValueListPacked is an alternative to the original UnmarshalValueList
```
Comment (aanthony1243) [inline services/collectd/service.go]: done.
